src/util/hash.rs

```diff
@@ -25,48 +25,50 @@ use core::cmp::min;
 use hashes::Hash;
 use consensus::encode::Encodable;
 
-/// Calculates the merkle root of a list of hashes inline
-/// into the allocated slice.
+/// Calculates the merkle root of a list of hashes inline (in place) in `data`.
 ///
 /// In most cases, you'll want to use [bitcoin_merkle_root] instead.
+///
+/// # Panics
+/// Panics if, and only if, `data` is empty. This function should only ever be called with the
+/// transaction data from a block, which should never be empty because blocks always contain at
+/// least one transaction (the coinbase transaction).
 pub fn bitcoin_merkle_root_inline<T>(data: &mut [T]) -> T
     where T: Hash + Encodable,
           <T as Hash>::Engine: io::Write,
 {
-    // Base case
-    if data.is_empty() {
-        return Default::default();
-    }
-    if data.len() < 2 {
-        return T::from_inner(data[0].into_inner());
+    if data.len() == 0 {
+        panic!("merkle root of an empty tree is undefined");
     }
-    // Recursion
-    for idx in 0..((data.len() + 1) / 2) {
-        let idx1 = 2 * idx;
-        let idx2 = min(idx1 + 1, data.len() - 1);
-        let mut encoder = T::engine();
-        data[idx1].consensus_encode(&mut encoder).unwrap();
-        data[idx2].consensus_encode(&mut encoder).unwrap();
-        data[idx] = T::from_engine(encoder);
+
+    // If there is only a single hash then that hash is, by definition, the merkle root.
+    if data.len() == 1 {
+        return data[0];
     }
-    let half_len = data.len() / 2 + data.len() % 2;
-    bitcoin_merkle_root_inline(&mut data[0..half_len])
+
+    merkle_root_r(data)
```

Comment:
This can be improved even more by

```rust
match data.len() {
    O => panic!(),
    1 => data[1],
    2 => merkle_root_r(data),
}
```

Reply:
I agree in principle, assuming you meant to write
```
match data.len() {
    O => panic!(),
    1 => data[0],
    _ => merkle_root_r(data),
}
```
For these two functions, however I tend to think this is not as clean:
1. The whole point of this PR is to make the two functions look the same (makes it quicker/easier to verify they are the same).
2. The final match arm of `bitcoin_merkle_root` is not a simple function call so this match style results in an additional layer if indentation which makes the whole function look a bit cluttered in my opinion. Here it is for reference:
```
    match iter.len() {
        0 => panic!("merkle root of an empty tree is undefined"),
        // If there is only a single hash then that hash is, by definition, the merkle root.
        1 => return iter.next().expect("cannot fail, length == 1"),
        _ => {
            // We need a local copy to pass to `merkle_root_r`. It's more efficient to do the first loop of
            // processing as we make the copy instead of copying the whole iterator.
            let half_len = iter.len() / 2 + iter.len() % 2;
            let mut alloc = Vec::with_capacity(half_len);
            while let Some(hash1) = iter.next() {
                // If the size is odd, use the last element twice.
                let hash2 = iter.next().unwrap_or(hash1);
                let mut encoder = T::engine();
                hash1.consensus_encode(&mut encoder).unwrap();
                hash2.consensus_encode(&mut encoder).unwrap();
                alloc.push(T::from_engine(encoder));
            }

            merkle_root_r(&mut alloc)
        }
    }
```

Thoughts?

Reply:
I deleted my original thoughts because they turned out to be wrong and were really long. PR now includes your suggestion.